### PR TITLE
EthereumRuntime: Use `EXTCODECOPY` to retrieve bytecode

### PR DESCRIPTION
**Pull request by pinkiebell**
Introduces the EVMCode helper library

Related #18

### CODE REVIEW

File: contracts/EVMCode.slb [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        address codeAddress;
+        bytes codeBytes;
+        uint length;
+    }
+
+    function fromAddress(address codeAddress) internal view returns (Code memory code) {
+        code.codeAddress = codeAddress;
+        uint codeSize;
+        assembly {
+            codeSize := extcodesize(codeAddress)
```
Comment (peara) [inline contracts/EVMCode.slb]: why don't you assign `code.length` directly here :thinking:

File: contracts/EVMCode.slb [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+    function toBytes(Code memory self) internal view returns (bytes memory bts) {
+        address codeContractAddress = self.codeAddress;
+
+        if (codeContractAddress == address(0)) {
+            bts = self.codeBytes;
+        } else {
+            uint size = self.length;
+            assembly {
+                bts := mload(0x40)
+                mstore(0x40, add(bts, and(add(add(size, 0x20), 0x1f), not(0x1f))))
```
Comment (peara) [inline contracts/EVMCode.slb]: please help me understand this line :smile:
I see you are allocate memory for code
what is the meaning of the `and( ... )` part?
And I think this line is worthy for some comment :thought_balloon:

File: contracts/Verifier.sol [hunk 3]
```diff
@@ -197,47 +197,23 @@ contract Verifier is EVMConstants, Ownable {
...
-        }
 
-        bytes memory code = new bytes(codeSize);
-        assembly {
-            extcodecopy(codeContractAddress, add(code, 0x20), codeptr, codeSize)
-        }
-
-        if ((dispute.state & END_OF_EXECUTION) != 0) {
-            if (uint8(code[_executionState.pc]) != OP_REVERT && uint8(code[_executionState.pc]) != OP_RETURN) {
+            if (opcode != OP_REVERT && opcode != OP_RETURN) {
```
Comment (peara) [inline contracts/Verifier.sol]: This question is unrelated to the PR.
Now I think more about this, I wonder if the operator in plasma chain would submit reverted transactions to main net?
Is this will open a kind of DOS attack?

File: contracts/EVMCode.slb (hunk 2 above)
Comment (pinkiebell) [reply contracts/EVMCode.slb]: Ya, that is padding up to a whole word

File: contracts/Verifier.sol (hunk 3 above)
Comment (pinkiebell) [reply contracts/Verifier.sol]: This can be challenged. This just makes sure that a challenger can proof that the solvers solution was too 'long'

File: contracts/EVMCode.slb (hunk 1 above)
Comment (pinkiebell) [reply contracts/EVMCode.slb]: Good question; Going to fix that 👍

Comment (pinkiebell) [reply contracts/EVMCode.slb]: Ha, I always forget this: That isn't possibly in an assembly block to (directly) assign a struct member (because memory)

Comment (peara) [reply contracts/EVMCode.slb]: I think you can do 
`mstore(add(code, 0x60), extcodesize(codeAddress))`
but that's look not very nice :laughing:

File: contracts/Verifier.sol (hunk 3 above)
Comment (peara) [reply contracts/Verifier.sol]: I see, this is just for the case solver didn't stop after hitting REVERT. My question was totally unrelated.
But given that, then stateHash is already enough to cover this case, isn't it?
Because if opcode is REVERT, EVMRuntime will return some error?

Comment (peara) [reply contracts/Verifier.sol]: Also there is opcode `invalid`. Should it be included here?

File: contracts/EVMCode.slb (hunk 1 above)
Comment (pinkiebell) [reply contracts/EVMCode.slb]: Yes, I wanted to do that at the second thought but didn't.
It is easy to forget about that and having strange errors if someone changes the structs.
We could make the offset a constant of course and putting this directly to the struct with a warning comment. Thoughts?

Comment (pinkiebell) [reply contracts/EVMCode.slb]: solc needs a `offsetof` macro 😁

File: contracts/Verifier.sol (hunk 3 above)
Comment (pinkiebell) [reply contracts/Verifier.sol]: Yes, stateHash is enough because they have to proof the execution and also agree on prioer execution steps.

>> Also there is opcode invalid. Should it be included here?

Hmm, no. Every valid execution ends with either RETURN or REVERT. 
Or most of the time: Out Of Gas xD

Comment (peara) [reply contracts/Verifier.sol]: I think so too. When something invalid happen it will REVERT.
But here: https://solidity.readthedocs.io/en/v0.5.2/assembly.html#opcodes
there is an `invalid` which also said to `end execution with invalid instruction`
Am I miss anything...

Comment (peara) [reply contracts/Verifier.sol]: I think `assert` will end with `invalid` but haven't found something conclusive

Comment (pinkiebell) [reply contracts/Verifier.sol]: @peara 
If we want to allow this, we might also include SELFDESTRUCT. But that wasn't at the table.
I think we don't want to allow that for PLASMA @johannbarbie ?

Comment (pinkiebell) [reply contracts/Verifier.sol]: @peara But still pointer, STOP is missing 😎

File: contracts/EVMCode.slb (hunk 1 above)
Comment (peara) [reply contracts/EVMCode.slb]: let's skip this at the moment :)

**Comment by pinkiebell** [comment]
Yay or Nay ? 🙂

**Comment by peara** [comment]
Yay :smile:
